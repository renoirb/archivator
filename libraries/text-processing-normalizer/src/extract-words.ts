import { wordNormalizer, nonStopWordIsser } from './predicates'

/**
 * Extract words and calculate usage frequency.
 *
 * ----
 *
 * See earlier implementation:
 *   v1.0.0, initial implementation:
 *     links:
 *       - name: extractWords
 *         url: https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L22-L46
 *       - name: analyze
 *         url: https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L76-L94
 *       - name: sort
 *         url: https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L63-L74
 *       - name: removePunctuation
 *         url: https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L18-L20
 *   v2.0.0, rework implementation:
 *     links:
 *       - name: normalize
 *         replacing: removePunctuation
 *         replacedWith: wordNormalizer
 *         url: https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L13-L16
 *       - name: extractWords:
 *         replacing: extractWords
 *         url: https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L18-L48
 *       - name: analyze
 *         replacing: analyze
 *         url: https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L50-L62
 *       - name: sort
 *         replacing: sort
 *         url: https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L64-L77
 */

export const extractWords = (
  body: string,
  stopWords: string[] = [],
  locales: string | string[] = ['en-CA', 'fr-CA'],
): Record<string, number> => {
  const normalizeWord = wordNormalizer(locales)
  const isNonStopWord = nonStopWordIsser(stopWords)
  // We want to use Object.create(null) instead of Map
  // because we want to quickly increment values on an Hash-Map
  const words: Record<string, number> = Object.create(null)
  const foundOnce = new Set<string>()
  const text = body.split(' ')
  for (let i = 0; i < text.length; i++) {
    const word = normalizeWord(text[i])
    if (word && isNonStopWord(word)) {
      if (foundOnce.has(word)) {
        if (Object.prototype.hasOwnProperty.call(words, word)) {
          words[word]++
        } else {
          words[word] = 2
        }
      } else {
        foundOnce.add(word)
      }
    }
  }
  return Object.seal(words)
}

// /**
//  * Take a collection of words,
//  * Return the same collection, sorted by usage.
//  *
//  * @param subject
//  */
// export function sort(subject) {
//   let sortable = []
//   for (let key in subject) {
//     sortable.push([key, subject[key]])
//   }
//   // Sort from more occurences, to least
//   sortable.sort((a, b) => {
//     return -1 * (a[1] - b[1])
//   })
//
//   return sortable // array in format [ [ key1, val1 ], [ key2, val2 ], ... ]
// }
//
// export const analyze = (words) => {
//   const keywords = new Map<string, number>()
//   const sorted = sort(words)
//   const max = 10
//   let iter = 0
//   for (let popular of sorted) {
//     let used = popular[1] // word has been used n times
//     let word = popular[0]
//     if (iter <= max && used > 3) {
//       keywords.set(word, used)
//     }
//     iter++
//   }
//
//   recv.data.keywords = keywords
//
//   return recv
// }
