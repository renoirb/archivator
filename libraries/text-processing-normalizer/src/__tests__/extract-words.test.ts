/* eslint-disable @typescript-eslint/ban-ts-ignore */

import { extractWords } from '../extract-words'

import { fixtures } from '.'

describe('extractWords', () => {
  test('Happy-Path', () => {
    const stopWords = fixtures.loadJson<ReadonlyArray<string>>(
      'english-stop-words-alpha.json',
    )
    const textContent = fixtures.loadText('article-alpha.txt')
    const subject = extractWords(textContent, [...stopWords])
    for (const w of stopWords) {
      expect(subject).not.toHaveProperty(w)
    }
    expect(subject).toMatchSnapshot()
  })
})
