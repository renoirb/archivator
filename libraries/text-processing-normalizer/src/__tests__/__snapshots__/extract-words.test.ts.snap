// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractWords Happy-Path 1`] = `
Object {
  "0": 2,
  "10101011": 2,
  "200": 2,
  "203011311": 2,
  "80": 2,
  "_call_awslatestmetadatas": 2,
  "_call_awsurl": 3,
  "_get_dreamcompute_additional": 2,
  "_get_dreamcompute_hostinfopath": 2,
  "_grains": 2,
  "_snake_caseify_dictdata": 2,
  "a_grainsdreamcomputepy": 2,
  "all": 4,
  "api": 39,
  "app1": 2,
  "availability_zone": 2,
  "available": 3,
  "aws": 4,
  "but": 3,
  "call_response_data": 4,
  "case": 3,
  "cluster": 6,
  "conn": 2,
  "connrequestget": 2,
  "d": 2,
  "data": 21,
  "def": 8,
  "dline": 3,
  "does": 2,
  "doesnt": 2,
  "dreamcompute": 5,
  "dreamcompute_info": 2,
  "dreamcomputeopenstack": 2,
  "each": 2,
  "ec2": 7,
  "elif": 2,
  "else": 4,
  "error": 2,
  "except": 5,
  "file": 5,
  "folder": 2,
  "following": 2,
  "for": 6,
  "from": 6,
  "function": 2,
  "get": 5,
  "grain": 3,
  "grains": 10,
  "grainsget": 3,
  "have": 2,
  "hostname": 4,
  "http": 2,
  "httplib": 2,
  "httplibhttpconnection169254169254": 2,
  "i": 6,
  "iad1": 3,
  "if": 7,
  "import": 4,
  "in": 15,
  "information": 4,
  "instance": 3,
  "internal": 2,
  "ioerror": 2,
  "it": 4,
  "key": 4,
  "keys": 2,
  "keys_to_mute": 3,
  "know": 2,
  "latestmetadata": 2,
  "latestmetadatablockdevicemappingamiinfo": 2,
  "latestmetadatablockdevicemappingebs0info": 2,
  "latestmetadatablockdevicemappingebs1info": 2,
  "latestmetadatablockdevicemappinginfo": 2,
  "latestmetadatablockdevicemappingrootinfo": 2,
  "latestmetadatahostnameinfo": 2,
  "latestmetadatainfo": 2,
  "latestmetadatainstanceactioninfo": 2,
  "latestmetadatainstanceidinfo": 2,
  "latestmetadatainstancetypeinfo": 2,
  "latestmetadatalocalipv4info": 2,
  "latestmetadataplacementavailabilityzoneinfo": 2,
  "latestmetadataplacementinfo": 2,
  "latestmetadatapublicipv4info": 2,
  "latestmetadataramdiskidinfo": 2,
  "latestmetadatareservationidinfo": 2,
  "lightspeed": 2,
  "like": 2,
  "line": 8,
  "log": 2,
  "logging": 2,
  "loginfocould": 3,
  "make": 3,
  "metadata": 7,
  "might": 2,
  "minion": 3,
  "more": 2,
  "my": 3,
  "name": 3,
  "nd": 2,
  "network_config": 2,
  "not": 10,
  "nova": 2,
  "on": 4,
  "openstack": 6,
  "openstack20120810meta_datajsoninfo": 2,
  "original": 5,
  "path": 5,
  "placement": 2,
  "public": 2,
  "publickeys": 3,
  "python": 3,
  "read": 5,
  "recursive": 2,
  "return": 13,
  "returns": 2,
  "s": 3,
  "salt": 8,
  "saltnovalocal": 3,
  "script": 8,
  "securitygroups": 2,
  "serr": 6,
  "service": 2,
  "should": 2,
  "some": 6,
  "sshrsa": 2,
  "stack": 2,
  "such": 2,
  "that": 17,
  "the": 37,
  "this": 6,
  "timeout1": 2,
  "to": 58,
  "try": 2,
  "url": 3,
  "v": 2,
  "vda": 2,
  "vm": 2,
  "was": 3,
  "what": 4,
  "which": 2,
  "will": 3,
  "within": 2,
  "work": 3,
  "x": 2,
  "you": 8,
  "your": 3,
}
`;
